File: leetcode3/최원준/3591. Check if Any Element Has Prime Frequency.py
#

'''
1. 아이디어 :
prime을 구하는법:
- 1 이하면 prime이 아니고,
- 2 면 prime이고,
- 2로 나뉘면 prime이고,
- 3부터 num의 제곱근까지 순회하면서 num 0으로 나누어 떨어지면 False
- 그 외 True

2. 시간복잡도 :
    O(n + n)

3. 자료구조/알고리즘 :
해시맵

'''

class Solution:
    def checkPrimeFrequency(self, nums: List[int]) -> bool:
        def is_prime(num):
            if num < 2:
                return False

            if num == 2:
                return True

            if num % 2 == 0:
                return False

            for i in range(3, int(num ** 0.5) + 1, 2):
                if num % i == 0:
                    return False

            return True

        counts = Counter(nums)
        for num, freq in counts.items():
            if is_prime(freq):
                return True
        return False

File: leetcode3/이진희/3591_Check if Any Element Has Prime Frequency.java
/*

1. 아이디어 : 에라토스테네스의 체와 빈도수 배열 사용
              100 까지 의 수므로 미리 각 숫자의 등장 빈도수와 100 이하의 소수를 구한다
              이후 빈도수 배열을 완전탐색해 소수가 존재하는지 찾는다

2. 시간복잡도 : O(N) + O(M log log M) + O(N) => O(N) (M=100)

3. 자료구조/알고리즘 : 에라토스테네스의 체, 빈도수 배열

 */

class Solution {
    public boolean checkPrimeFrequency(int[] nums) {
        // 특정 수의 등장 횟수가 소수면 return true

        int[] cnt = new int[101];
        for(int i : nums) {
            cnt[i]++;
        }

        boolean[] isNotPrime = new boolean[101];
        isNotPrime[0] = isNotPrime[1] = true;

        for(int i=2; i*i<101; i++) {
            for(int j=i*i; j<101; j+=i) {
                isNotPrime[j] = true;
            }
        }

        for(int i : cnt) {
            if(isNotPrime[i]) continue;
            return true;
        }
        return false;
    }
}